### PR TITLE
Bug-5145

### PR DESCRIPTION
**Pull request by jroy1992**
Added new setting Node Classes To Exclude to exclude nodes belonging to particular node class from user file path validation
Also, for duplicate path on write node validation, making sure we don’t treat the same node as duplicate path.

### CODE REVIEW

**Comment by jroy1992** [comment]
In combinatin with http://dd-git.d2.com/locations/india/sgtk_config/merge_requests/79

File: hooks/tk-multi-publish2/nuke/publish.py [hunk 1]
```diff
@@ -548,17 +560,19 @@ def _write_node_path_duplicacy(self, item):
...
                                    if value == node_path]
-            self.logger.error("Duplicate output path.",
-                              extra={
-                                  "action_show_more_info": {
-                                      "label": "Show Info",
-                                      "tooltip": "Show node(s) with identical output path",
-                                      "text": "Following node(s) have same output path as {}:\n\n{}".
-                                      format(node_name, '\n'.join(duplicate_path_node))
+
+            if node_name not in duplicate_path_node:
```
Comment (aparna-d2) [inline hooks/tk-multi-publish2/nuke/publish.py]: This seems risky. Even in node_name is present in the list, we could have other nodes which have duplicate paths, which should still make validation return False.
Perhaps we can delete the node_name if it exists, and then check whether there are any nodes left in the list?
Also, maybe rename `duplicate_path_node` to `duplicate_path_nodes`.

File: hooks/tk-multi-publish2/nuke/publish.py [hunk 2]
```diff
@@ -548,17 +560,19 @@ def _write_node_path_duplicacy(self, item):
         if node_path in all_paths:
             duplicate_path_node = [key for (key, value) in item.parent.properties['write_node_paths_dict'].items()
                                    if value == node_path]
```
Comment (aparna-d2) [inline hooks/tk-multi-publish2/nuke/publish.py]: If the fetch pattern is by node path, maybe the dict should be organised as `{"path1": ["node1", "node2"]}`? Then we can just ask for `dict[node_path]`
If it is also being accessed by node_name, feel free to ignore this.

File: hooks/tk-multi-publish2/nuke/publish.py (hunk 1 above)
Comment (jroy1992) [reply hooks/tk-multi-publish2/nuke/publish.py]: The logic here is that each unique path will be associated with the first node having that path. If there are other nodes having the same path, then they will be compared against that first node. That's why the logic `if node_name not in duplicate_path_node:`. 
The error message is a bit misleading. Should have fixed that.

But yea its better to show list of all nodes sharing the path. Fixing the logic.

**Comment by jroy1992** [comment]
Added a post_phase hook to reset `visited_dict` and `write_node_paths_dict` properties on Nuke Session.

File: hooks/tk-multi-publish2/nuke/publish.py [hunk 3]
```diff
@@ -556,24 +556,27 @@ def _non_sgtk_writes(self):
     def _write_node_path_duplicacy(self, item):
         node_path = item.properties['node']['cached_path'].value()
         node_name = item.properties['node'].name()
-        all_paths = item.parent.properties['write_node_paths_dict'].values()
+        all_paths = item.parent.properties['write_node_paths_dict'].keys()
         if node_path in all_paths:
-            duplicate_path_node = [key for (key, value) in item.parent.properties['write_node_paths_dict'].items()
-                                   if value == node_path]
+            duplicate_path_nodes = item.parent.properties['write_node_paths_dict'].get(node_path, [])
```
Comment (aparna-d2) [inline hooks/tk-multi-publish2/nuke/publish.py]: Default value should be `set()`, to allow `add()` and to remain consistent.

File: hooks/tk-multi-publish2/nuke/publish.py [hunk 4]
```diff
@@ -556,24 +556,27 @@ def _non_sgtk_writes(self):
...
-                                          format(node_name, '\n'.join(duplicate_path_node))
+                                          format(node_name, '\n'.join(duplicate_path_nodes))
                                       }
                                   }
                                   )
+                item.parent.properties['write_node_paths_dict'][node_path].add(node_name)
                 return False
-        item.parent.properties['write_node_paths_dict'] = {node_name: node_path}
+
+        item.parent.properties['write_node_paths_dict'].update({node_path: set([node_name])})
```
Comment (aparna-d2) [inline hooks/tk-multi-publish2/nuke/publish.py]: Prefer using `dict[key] = value` over `dict.update({key: val})` as we unnecessarily create an extra dict in the latter method. Also, performance justification: https://stackoverflow.com/questions/15456158/python-dict-update-vs-subscript-to-add-a-single-key-value-pair
Also, we can just say `{node_name}` to create a set, thus avoiding a list creation too.

**Pull request merged**